Themes: Use new Redux state for active theme information

ockham opened the pull request:
Use the new reducer, selectors, actions, and query components (instead of the old current-theme Redux subtree) to keep track of a site's currently active theme. This requires enabling a number of new reducers, including the `queries` one, which we need to store themes fetched by `<QueryTheme />` via `requestTheme`.

_Conveniently fixes #9379_ :smile:

To test: Verify that the theme showcase works as before, with a special focus on theme activation (in different modes). Also verify that #9379 is fixed.

_Edit:_ Additionally required testing: Verify that all ways of activating a theme still work (and cover different modes, i.e. single-site, multi-site, etc):

'...' menu and theme sheets;
via 'Try & Customize' -> 'Save & Activate'
via purchase

ockham commented on client/components/track-interactions/README.md:
```diff
@@ -48,9 +48,9 @@ clicking yields:
-const mapPropsToAction = props => ( {
-	name: props.theme.id,
-	isCurrentTheme: props.theme.id === getCurrentTheme( state ),
+const mapPropsToAction = ( state, { theme } ) => ( {
```
@mcsf I've made `state` explicit here b/c `getCurrentTheme` was using it ^ (and now `isActiveTheme` is), but looking at https://github.com/Automattic/wp-calypso/blob/master/client/components/track-interactions/index.js#L53 it looks like we aren't even passing `state` in. So... what do we do about this example? 😄

seear commented on client/my-sites/themes/current-theme/index.jsx:
```diff
@@ -88,7 +93,13 @@ const CurrentThemeWithOptions = ( { siteId, currentTheme } ) => (
+		const siteIdOrWpcom = isJetpackSite( state, siteId ) ? siteId : 'wpcom';
+		return {
+			currentThemeId,
+			currentTheme: getTheme( state, siteIdOrWpcom, currentThemeId ),
```
Probably a bit late with this, but would it make sense to have `siteId` as the third parameter to `getTheme()`, defaulting to `'wpcom'`?

ockham replied:
Yeah, that makes sense. We'd have to keep other selectors and their signatures in mind for consistency; and I'd rather tackle it in a separate PR.

seear replied:
> We'd have to keep other selectors and their signatures in mind for consistency

Yeah that's the trade-off, so it may not make sense.

> I'd rather tackle it in a separate PR.

Absolutely

folletto commented:
Tested on Chrome:

* Text search — ✅ 
* Filter search — ✅ 
* All / Free / Premium — ✅ 
* Switch single site to multi site — ☑️  (no regression)
* Activate (single site) from ••• — ✅ 
* Activate (single site) from sheet — 🔴 
* Activate (multi site) from ••• — ✅ 
* Activate (multi site) from sheet — 🔴 
* Bug #9379 (no author on activation) fixed — ✅  

It seems that activation from the sheet gets stuck here:

![screen shot 2016-11-30 at 13 50 28](https://cloud.githubusercontent.com/assets/4389/20754694/0c6e7e86-b704-11e6-9c42-d36e55cf466f.png)

It appears that after "Activating site..." it gets stuck. No error in console.

seear commented:
I'm getting `Failed prop type` (value undefined) warnings for these in ThanksModal:
```
isActivating: React.PropTypes.bool.isRequired,
 hasActivated: React.PropTypes.bool.isRequired,
```

seear commented:
Seeing this after purchase from multi-site theme sheet, on a site that already has the theme:

<img width="1273" alt="screen shot 2016-11-30 at 14 43 05" src="https://cloud.githubusercontent.com/assets/7767559/20756822/130540b0-b70c-11e6-84aa-3609182d521e.png">


<img width="1007" alt="screen shot 2016-11-30 at 14 43 46" src="https://cloud.githubusercontent.com/assets/7767559/20756681/81bd9ca6-b70b-11e6-958a-eba8f3cc4230.png">

`master` handles the error properly:
<img width="864" alt="screen shot 2016-11-30 at 14 47 24" src="https://cloud.githubusercontent.com/assets/7767559/20756808/0a37a6e4-b70c-11e6-8c98-87383b89ea11.png">

seear commented:
> Seeing this after purchase from multi-site theme sheet, on a site that already has the theme:

Ok, ignore that. Seems like I was testing without the latest commits on this branch. I see the same behaviour as `master` now.

😅

ockham commented:
> Ok, ignore that. Seems like I was testing without the latest commits on this branch. I see the same behaviour as `master` now.

I think you're actually right about the error; it might seem fixed because your Redux state has now some data in the relevant place, but I think I've found a piece of code that really needs fixing. Good catch 😄

ockham commented:
Okay, I reworked the way we use `themeActivated`, see lengthy commit msg of one of the lastet commits in here. Tried to split off into separate PR (#9740) but unfortunately that doesn't work without putting in some backporting work, which isn't worth it. So we'll have to review and test as part of this PR. (I added some additional testing instructions to the PR desc.)

seear commented on client/state/themes/actions.js:
```diff
@@ -362,7 +362,8 @@ export function activateTheme( themeId, siteId, source = 'unknown', purchased =
 		return wpcom.undocumented().activateTheme( themeId, siteId )
 			.then( ( theme ) => {
-				dispatch( themeActivated( theme, siteId, source, purchased ) );
+				const themeStylesheet = theme.stylesheet || themeId; // Fall back to ID for Jetpack sites which don't return a stylesheet attr.
```
Will falling back to ID for jetpack sites cause a problem for sites state?

ockham replied:
No, for Jetpack sites, sites state (`options.theme_slug`) uses the un-prefixed theme ID, too.

seear commented on client/my-sites/customize/actions.js:
```diff
@@ -33,18 +33,12 @@ var CustomizeActions = {
 
-		themeActivated( id, site.ID, 'customizer' );
-
-		Dispatcher.handleViewAction( {
```
Nice!

seear commented:
Jetpack activate works but shows nothing in thanks modal:
<img width="521" alt="screen shot 2016-12-01 at 12 53 47" src="https://cloud.githubusercontent.com/assets/7767559/20794673/4f5cbfac-b7c5-11e6-939b-1389666679f6.png">

ockham commented:
Latest commit should've fixed thanks modal (and current theme bar). Be sure to test both JP and WPCOM sites now to ensure I didn't introduce a regression with the latter :-p

folletto commented:
Tested on Chrome:

* Text search — ✅ 
* Filter search — ✅ 
* All / Free / Premium — ✅ 
* Switch single site to multi site — ☑️  (no regression)
* Activate (single site) from ••• —  ✅ 
* Activate (single site) from sheet — ✅ 
* Activate (multi site) from ••• — ✅ 
* Activate (multi site) from sheet — ✅ 
* Activate from customizer (Try) — ✅ 
* Bug #9379 (no author on activation) fixed — ✅

seear commented:
Code looks good 👍 

Will test now.

seear commented:
Jetpack sites work fine. Combined with testing above from @folletto I'd say it is time to merge.
